NeuroID/src/main/java/com/neuroid/tracker/service/NIDJobServiceManager.kt

```diff
@@ -48,61 +51,110 @@ class NIDJobServiceManager(
         this.userActive = true
         this.clientKey = clientKey
         this.application = application
-        jobCaptureEvents = createJobServer()
+
+        sendEventsJob = createSendEventsServer()
+        sendCadenceJob = createSendCadenceServer()
         NIDSensorHelper.initSensorHelper(application, logger)
 
-        gyroCadenceJob = createGyroJobServer()
+        gyroCadenceJob = createGyroCadenceServer()
         activityManager = application.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
         this.isSetup = true
     }
 
     @Synchronized
     fun stopJob() {
         NIDSensorHelper.stopSensors()
-        jobCaptureEvents?.cancel()
-        jobCaptureEvents = null
+        sendCadenceJob?.cancel()
+        sendCadenceJob = null
 
         gyroCadenceJob?.cancel()
         gyroCadenceJob = null
     }
 
     @Synchronized
     fun isStopped(): Boolean {
-        return jobCaptureEvents?.isActive != true
+        return sendCadenceJob?.isActive != true
     }
 
     @Synchronized
     fun restart() {
         this.userActive = true
         NIDSensorHelper.restartSensors()
-        jobCaptureEvents?.cancel()
-        jobCaptureEvents = createJobServer()
+
+        // If the job has ended for any reason restart it
+        if (!sendEventsJob.isActive) {
+            sendEventsJob.cancel()
+            sendEventsJob = createSendEventsServer()
+        }
+
+        sendCadenceJob?.cancel()
+        sendCadenceJob = createSendCadenceServer()
 
         gyroCadenceJob?.cancel()
-        gyroCadenceJob = createGyroJobServer()
+        gyroCadenceJob = createGyroCadenceServer()
+    }
+
+    /**
+     * Notify the send events server to initiate a send.
+     */
+    fun sendEvents(forceSendEvents: Boolean = false) {
+        CoroutineScope(dispatcher).launch {
+            sendEventsNotification.send(forceSendEvents)
+        }
+    }
+
+    /**
+     * Creates a job that synchronizes all requests to send events.
+     */
+    private fun createSendEventsServer(): Job {
+        return CoroutineScope(dispatcher).launch {
+            try {
+                for (notification in sendEventsNotification) {
+                    var sendNow = notification
+                    var next = sendEventsNotification.tryReceive()
+                    while (next.isSuccess) {
+                        next.getOrNull()?.let {
+                            sendNow = sendNow || it
+                        }
+                        next = sendEventsNotification.tryReceive()
+                    }
```

Comment:
Trying to understand why we need the while loop and tryReceive()-ing the next item when we are already iterating in the notification queue for all known values. The next 5 second iteration will pick up the items in the queue after the current iteration. It looks like we want to get all successful entries during the time that we process the items already in the queue?

Reply:
Great question, the idea is that if we have called sendEvents multiple times (lets say 5 times) within the 5 seconds either due to the 5 seconds being hit or the forceSend parameter, we don't actually want to make 5 requests, so instead we do a loop to combine all the requests into one and only make one request to our collection service to be more efficient

Reply:
Whoever wrote this code should have commented it.  Can you correct that mistake?